cpp/server/proxy.cc

```diff
@@ -0,0 +1,134 @@
+
+
+void ProxyRequestDone(libevent::Base* base, JsonOutput* output,
+                      evhttp_request* request, const std::string& path,
+                      UrlFetcher::Response* response, util::Task* task) {
+  CHECK_NOTNULL(request);
+  CHECK_NOTNULL(response);
+  CHECK_NOTNULL(task);
+  total_proxied_requests->Increment(path);
+  total_proxied_responses->Increment(path, response->status_code);
+
+  if (!task->status().ok()) {
+    return output->SendError(request, HTTP_INTERNAL,
+                             "Proxied request failed.");
+  }
+
+  // TODO(alcutter): Consider retrying the proxied request some number of times
+  // in the case where the request fails.
+  for (auto it(response->headers.begin()); it != response->headers.end();
+       ++it) {
+    CHECK_EQ(evhttp_add_header(evhttp_request_get_output_headers(request),
+                               it->first.c_str(), it->second.c_str()),
+             0);
+  }
+  CHECK_GT(evbuffer_add_printf(evhttp_request_get_output_buffer(request), "%s",
+                               response->body.c_str()),
+           -1);
+
+  const int response_code(response->status_code);
+  base->Add([request, response_code]() {
+    evhttp_send_reply(request, response_code, /*reason*/ NULL,
+                      /*databuf*/ NULL);
+  });
+}
+
+
+}  // namespace
+
+
+Proxy::Proxy(libevent::Base* base, JsonOutput* output,
+             const GetFreshNodesFunction& get_fresh_nodes, UrlFetcher* fetcher,
+             util::Executor* executor)
+    : base_(base),
+      output_(output),
+      get_fresh_nodes_(get_fresh_nodes),
+      fetcher_(fetcher),
+      executor_(executor) {
+}
+
+
+void Proxy::ProxyRequest(evhttp_request* req) const {
+  CHECK_NOTNULL(req);
+
+  const std::vector<ct::ClusterNodeState> fresh_nodes(get_fresh_nodes_());
+  if (fresh_nodes.empty()) {
+    return output_->SendError(req, HTTP_SERVUNAVAIL,
+                              "No node able to serve request.");
+  }
+  const ct::ClusterNodeState& target(
+      fresh_nodes[std::rand() % fresh_nodes.size()]);
+
+  URL url(evhttp_request_uri(req));
+  url.SetProtocol("http");
+  url.SetHost(target.hostname());
+  url.SetPort(target.log_port());
+
+  UrlFetcher::Request fetcher_req(url);
+
+  switch (evhttp_request_get_command(req)) {
+    case EVHTTP_REQ_DELETE:
+      fetcher_req.verb = UrlFetcher::Verb::DELETE;
+      break;
+    case EVHTTP_REQ_GET:
+      fetcher_req.verb = UrlFetcher::Verb::GET;
+      break;
+    case EVHTTP_REQ_POST:
+      fetcher_req.verb = UrlFetcher::Verb::POST;
+      break;
+    case EVHTTP_REQ_PUT:
+      fetcher_req.verb = UrlFetcher::Verb::PUT;
+      break;
+    default:
+      return output_->SendError(req, HTTP_BADMETHOD, "Bad method requested.");
+      break;
+  }
+
+  for (evkeyval* ptr = evhttp_request_get_input_headers(req)->tqh_first; ptr;
+       ptr = ptr->next.tqe_next) {
+    fetcher_req.headers.insert(std::make_pair(ptr->key, ptr->value));
```

Comment:
Again, shouldn't copy *all* the headers, otherwise problems... Should be able to make a helper function for that.